src/app/tests/TestAttributeCache.cpp

```diff
@@ -549,9 +549,13 @@ nlTestSuite theSuite =
 
 int TestAttributeCache()
 {
-    TestContext gContext;
-    gSuite = &theSuite;
-    nlTestRunner(&theSuite, &gContext);
+    VerifyOrReturnError(chip::Platform::MemoryInit() == CHIP_NO_ERROR, FAILURE);
```

Comment:
Why? `TestContext::Initialize` (which is the suite initalizer) calls `LoopbackMessagingContext::Initialize` here, which calls `LoopbackMessagingContext::Init`, which calls `Platform::MemoryInit`.

Is there something in the TestContext constructor that ends up needing Platform::Memory bits?

This applies to a lot of these test changes: the objects we are using are supposed to be doing MemoryInit themselves already....

Reply:
Some thing(s) within a TestContext end up freeing pool objects on destruction ([one example](https://github.com/project-chip/connectedhomeip/blob/245035ccc200b6afba9a7020beba325c5c7e8563/src/transport/GroupSession.h#L76)), which would be after `Platform::MemoryShutdown`. (I think the whole `Platform::Memory` approach is wrong, but that's another story.) So the test changes are expedient but hopefully not permanent — I think in general we do want CHIP classes to be statically allocatable.

Since inline vs heap allocation of pool objects is supposed to be transparent to the pool client, anything that fails when `BitMapObjectPool` is changed to a heap `ObjectPool` strongly suggests a lifetime problem somewhere, and heap objects also make ASAN checking easier (although it's apparently [possible](https://github.com/google/sanitizers/wiki/AddressSanitizerManualPoisoning) to improve coverage for `BitMapObjectPool` as well).

Reply:
How about we give LoopbackMessagingContext a superclass which has the following behaviors:
1. Has a boolean member indicating whether its init succeeded.
2. Has an Init method that calls Platform::MemoryInit and if that succeeds sets the boolean.
3. Has a destructor that calls Platform::MemoryShutdown.  Notably, this would run after ~LoopbackMessagingContext, so we would get the right scoping.

We can then probably remove the existing MemoryInit/MemoryShutdown from LoopbackMessagingContext::Init/Shutdown...

Thoughts?

Reply:
Sure. I had initially thought this PR would be a trivial replacement to confirm that pools are agnostic to the backing storage, but the unfortunate reality is that it's a method for detecting memory corruption, so it's worth more time on the tests.

Reply:
OK.  If you would prefer to change LoopbackMessagingContext in a followup I can also live with that... Just want to make sure we don't leave things in a state where writing a test is more complicated than it needs to be.